File: 211. Add and Search Word - Data structure design.py
class TrieNode:
    def __init__(self):
        self.child = {}
        self.isEndofWord = False

class WordDictionary:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        """
        Adds a word into the data structure.
        """
        node = self.root
        length = len(word)
        for level in range(length):
            pos = word[level]
            if pos not in node.child:
                node.child[pos] = TrieNode()
            node = node.child[pos]
        node.isEndofWord = True

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        """
        self.ret = False
        self.dfs(word, self.root)
        return self.ret
        
    def dfs(self, word, node):
        if not word:
            if node.isEndofWord:
                self.ret = True
            return
        
        if word[0] == '.':
            for i in node.child:
                self.dfs(word[1:], node.child[i])
        else:
            if word[0] in node.child:
                self.dfs(word[1:], node.child[word[0]])
            else:
                return

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: 211. Add and Search Word - Data structure design.java
class WordDictionary {
    class TrieNode {
        boolean isEnd;
        TrieNode[] children;
        public TrieNode() {
            isEnd = false;
            children = new TrieNode[26];
        }
    }

    TrieNode root;
    /** Initialize your data structure here. */
    public WordDictionary() {
        root = new TrieNode();
    }
    
    /** Adds a word into the data structure. */
    public void addWord(String word) {
        TrieNode curr = root;
        for (int i=0; i<word.length(); i++) {
            char c = word.charAt(i);
            if (curr.children[c-'a'] == null) {
                curr.children[c-'a'] = new TrieNode();
            }
            curr = curr.children[c-'a'];
        }
        curr.isEnd = true;
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    public boolean search(String word) {
        return search(word, 0, root);
    }
    
    private boolean search(String word, int pos, TrieNode node) {
        if (pos == word.length()) {
            return node.isEnd? true : false;
        }
        char c = word.charAt(pos);
        boolean res = false;
        if (c == '.') {
            for (TrieNode next : node.children) {
                if (next != null) {
                    res = res || search(word, pos+1, next);
                }
            }
        } else {
            if (node.children[c-'a'] == null) {
                res = false;
            } else {
                res = search(word, pos+1, node.children[c-'a']);
            }
        }
        return res;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */